service/history/decision/handler.go

```diff
@@ -480,7 +480,8 @@ Update_History_Loop:
 				tag.WorkflowID(token.WorkflowID),
 				tag.WorkflowRunID(token.RunID),
 				tag.WorkflowDomainID(domainID))
-			msBuilder, err = failDecisionHelper(ctx, wfContext, scheduleID, startedID, failCause, []byte(failMessage), request)
+			msBuilder, err = handler.failDecisionHelper(ctx, wfContext, scheduleID, startedID, failCause,
```

Comment:
If workflow is terminated here, do we need to exit the method and return an error like what we did for `TransactionSizeLimitError`? I am not sure if the rest of the logic here and in the frontend handler still work for terminated workflow.

Reply:
It's too early to terminate here since the mutable state needs to be updated (line 547). All other code is safe to execute since the state of the workflow transitions to Terminated. Having said that, the very last section in the function about dispatching activity/decision tasks is pointless if the workflow is no longer running. I'll add an early return there.